### PR TITLE
Infer optional types for IIFE parameters with missing arguments

### PR DESCRIPTION
**Pull request by ahejlsberg**
We currently contextually type un-annotated [IIFE](https://en.wikipedia.org/wiki/Immediately-invoked_function_expression) parameters by their corresponding arguments. This PR expands upon this concept by inferring un-annotated IIFE parameter with *missing* arguments as optional. That allows us to handle the following common idiom with no errors:

```ts
let foo = (function (window, undefined) {
    // Body
})(window);
```

Without this PR the example is an error because the `undefined` isn't marked optional.

Note that the PR also simplifies some overly complicated logic for computing the `minArgCount` property in signatures.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@billti @mhegazy This is a fix for an issue Bill mentioned the other day.

**Comment by billti** [comment]
Great!

Just playing around with this in TypeScript, and with `strictNullChecks` set, the below gives an error **with** the type annotation on `b`. Without the annotation on `b` it is fine (and also obviously if I pass `undefined` for the second arg).

Maybe this is "by design", and if you put the annotation you need to mark the param as optional (which also gets rid of the error), but intuitively it feels like this should be valid with the new inference. Thanks!

```typescript
(function(a: string, b: undefined) {

})("test");
```

**Comment by ahejlsberg** [comment]
@billti Yes, this is by design. We only contextually type un-annotated parameters. Once a type annotation is present you must also add a `?` modifier if you want the parameter to be considered optional.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10933,23 +10925,23 @@ namespace ts {
...
+                            restTypes.push(getWidenedLiteralType(checkExpression(iife.arguments[i])));
+                        }
+                        return restTypes.length ? createArrayType(getUnionType(restTypes)) : undefined;
+                    }
+                    const links = getNodeLinks(iife);
+                    const cached = links.resolvedSignature;
+                    links.resolvedSignature = anySignature;
+                    const type = indexOfParameter < iife.arguments.length ?
+                        getWidenedLiteralType(checkExpression(iife.arguments[indexOfParameter])) :
+                        parameter.initializer ? undefined : undefinedWideningType;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: @sandersn this will need to be covered in your change for initialized parameters.

**Pull request merged**